### PR TITLE
support multiple distros in SCT

### PR DESCRIPTION
**Pull request by amoskong**
Currently we only support to use centos AMI/gce instance/other backends, we want to test upgrade of other distros (ubuntu14.04/ubuntu16.04/debian8/...).

This PR added support of multi distros in BaseNode, and BaseScyllaCluster, let's try to wrapper distro gaps insider those two class.

Welcome your comments, thanks!

### CODE REVIEW

File: sdcm/cluster.py [hunk 1]
```diff
@@ -749,6 +777,9 @@ def get_backtraces(self):
         Verify the number of backtraces stored, report if new ones were found.
         """
         self.wait_ssh_up(verbose=False)
+        if self.distro == 'ubuntu14':
+            # ubuntu14 doesn't has coredumpctl, skip it.
```
Comment (bentsi) [inline sdcm/cluster.py]: http://manpages.ubuntu.com/manpages/xenial/man1/coredumpctl.1.html

File: sdcm/cluster.py [hunk 2]
```diff
@@ -1657,15 +1727,25 @@ def check_nodes_up_and_normal(self, nodes=[]):
 
     def get_scylla_version(self):
         if not self.nodes[0].scylla_version:
-            try:
-                result = self.nodes[0].remoter.run("rpm -q {}".format(self.nodes[0].scylla_pkg()))
-            except Exception as ex:
-                self.log.error('Failed getting scylla version: %s', ex)
+            if self.nodes[0].get_distro == 'centos':
```
Comment (bentsi) [inline sdcm/cluster.py]: self.nodes[0].distro

File: sdcm/cluster_aws.py [hunk 3]
```diff
@@ -417,9 +417,8 @@ def node_setup(self, node, verbose=False):
                                   server_encrypt=self._param_enabled('server_encrypt'),
                                   client_encrypt=self._param_enabled('client_encrypt'))
 
-            node.remoter.run('sudo yum install -y {}-gdb'.format(node.scylla_pkg()),
```
Comment (bentsi) [inline sdcm/cluster_aws.py]: Why did you remove scylla_pkg installation? Is it part of the image now?

File: sdcm/cluster.py [hunk 4]
```diff
@@ -322,15 +322,42 @@ def __init__(self, name, ssh_login_info=None, base_logdir=None, node_prefix=None
...
+        if 'Ubuntu 14.04' in result.stdout:
+            self.distro = 'ubuntu14'
+        elif 'Ubuntu 16.04' in self.stdout:
+            self.distro = 'ubuntu16'
+        elif 'Debian GNU/Linux 8' in result.stdout:
+            self.distro = 'debian8'
+        elif 'Debian GNU/Linux 9' in result.stdout:
+            self.distro = 'debian9'
+        else:
+            self.log.debug("fail to detect the distro name, %s" % result.stdout)
```
Comment (bentsi) [inline sdcm/cluster.py]: Failed to detect

File: sdcm/cluster.py [hunk 5]
```diff
@@ -322,15 +322,42 @@ def __init__(self, name, ssh_login_info=None, base_logdir=None, node_prefix=None
...
+        if self.distro:
+            return self.distro
+
+        result = self.remoter.run('grep CentOS /etc/redhat-release', ignore_status=True)
+        if result.exit_status == 0:
+            self.distro = 'centos'
+
+        result = self.remoter.run('cat /etc/issue', ignore_status=True)
+        if 'Ubuntu 14.04' in result.stdout:
+            self.distro = 'ubuntu14'
```
Comment (bentsi) [inline sdcm/cluster.py]: Generally you should avoid using strings since it is easy to make mistake and code completion will not work
I would use Enum instead Like OS.ubuntu14, OS.ubuntu16, etc.

File: sdcm/cluster.py [hunk 6]
```diff
@@ -322,15 +322,42 @@ def __init__(self, name, ssh_login_info=None, base_logdir=None, node_prefix=None
         self.is_enterprise = None
         self.replacement_node_ip = None  # if node is a replacement for a dead node, store dead node private ip here
         self._set_prometheus_paths()
+        self.distro = None
```
[the remark this answers is not in the document]
Comment (amoskong) [reply sdcm/cluster.py]: Current get_distro() requests instance network. If we get distro during init stage, it will wait there for a while. We expect multiple instances can't be created & setup in better parallel.

File: sdcm/cluster.py [hunk 7]
```diff
@@ -322,15 +322,42 @@ def __init__(self, name, ssh_login_info=None, base_logdir=None, node_prefix=None
...
 
     def is_docker(self):
         return self.__class__.__name__ == 'DockerNode'
 
     def scylla_pkg(self):
         if self.is_enterprise is None:
-            result = self.remoter.run("sudo yum search scylla-enterprise", ignore_status=True)
-            self.is_enterprise = True if ('scylla-enterprise.x86_64' in result.stdout or
-                                          'No matches found' not in result.stdout) else False
+            if self.get_distro() == 'centos':
```
[the remark this answers is not in the document]
Comment (amoskong) [reply sdcm/cluster.py]: I have idea to wrap helper function to install packages for multiple distros.
As you said, we also need function to judge the distro. I will fix it.

File: sdcm/cluster.py (hunk 2 above)
Comment (amoskong) [reply sdcm/cluster.py]: ;-) fixed locally

File: sdcm/cluster.py (hunk 4 above)
Comment (amoskong) [reply sdcm/cluster.py]: Ok.

File: sdcm/cluster_aws.py (hunk 3 above)
Comment (amoskong) [reply sdcm/cluster_aws.py]: It's just installing the scylla-gdb package, I moved the installation to `node_setup() -> config_setup()`

File: sdcm/cluster.py (hunk 5 above)
Comment (amoskong) [reply sdcm/cluster.py]: I prefer to use enum, but I found `enum` isn't installed by default in my machine.
So I didn't use that.

File: sdcm/cluster.py (hunk 1 above)
Comment (amoskong) [reply sdcm/cluster.py]: It's for 16.04

`http://manpages.ubuntu.com/manpages/trusty/en/man1/coredumpctl.1.html` isn't available

File: sdcm/cluster.py [hunk 8]
```diff
@@ -322,15 +323,103 @@ def __init__(self, name, ssh_login_info=None, base_logdir=None, node_prefix=None
...
+
+        return self.distro
+
+    def is_centos7(self):
+        self.distro == Distro.CENTOS7
+
+    def is_rhel7(self):
+        self.distro == Distro.RHEL7
+
+    def like_centos(self):
```
Comment (bentsi) [inline sdcm/cluster.py]: It is better to say `is_redhat_like` since CentOS is open source distribution that is RHEL compatible

File: sdcm/cluster.py [hunk 9]
```diff
@@ -1498,6 +1641,7 @@ def start_scylla(node, queue):
         self.log.debug('Update DB binary duration -> %s s', int(time_elapsed))
 
     def _update_db_packages(self, new_scylla_bin, node_list):
+        # fixme: only support centos distro
```
Comment (bentsi) [inline sdcm/cluster.py]: What does this #fixme mean?

File: sdcm/cluster.py [hunk 10]
```diff
@@ -1659,15 +1803,25 @@ def check_nodes_up_and_normal(self, nodes=[]):
 
     def get_scylla_version(self):
         if not self.nodes[0].scylla_version:
-            try:
-                result = self.nodes[0].remoter.run("rpm -q {}".format(self.nodes[0].scylla_pkg()))
-            except Exception as ex:
-                self.log.error('Failed getting scylla version: %s', ex)
+            if self.nodes[0].like_centos():
+                try:
+                    result = self.nodes[0].remoter.run("rpm -q {}".format(self.nodes[0].scylla_pkg()))
```
Comment (bentsi) [inline sdcm/cluster.py]: Can we use `scylla --version` here?

File: sdcm/cluster.py (hunk 9 above)
Comment (amoskong) [reply sdcm/cluster.py]: _update_db_packages() isn't the common usecase, I want to support other distros later.

File: sdcm/cluster.py (hunk 10 above)
Comment (amoskong) [reply sdcm/cluster.py]: Yes

File: sdcm/cluster.py (hunk 8 above)
Comment (amoskong) [reply sdcm/cluster.py]: I will use `is_rhel_like`

**Comment by roydahan** [comment]
@amoskong ping

**Comment by roydahan** [comment]
@bentsi / @larisau, please help with review, I would like to merge this ASAP.

File: sdcm/cluster.py [hunk 11]
```diff
@@ -2134,11 +2300,34 @@ def wait_for_init(self):
...
-            yum install -y python-pip unzip wget docker
-            pip install --upgrade pip
-            pip install pyyaml
-        """.format(**locals()))
+        if node.is_rhel_like():
+            prereqs_script = dedent("""
+                yum install -y python-pip unzip wget docker
+                pip install --upgrade pip
+                pip install pyyaml
+            """.format(**locals()))
```
Comment (bentsi) [inline sdcm/cluster.py]: Why do you use `.format` if there is nothing to format in the string?
Please remove `.format`

File: sdcm/cluster.py [hunk 12]
```diff
@@ -2134,11 +2300,34 @@ def wait_for_init(self):
...
+            prereqs_script = dedent("""
+                curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -
+                sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
+                sudo apt-get update
+                sudo apt-get install -y docker docker.io
+                apt-get install -y python-setuptools unzip wget
+                easy_install pip
+                pip install --upgrade pip
+                pip install pyyaml
+            """.format(**locals()))
```
Comment (bentsi) [inline sdcm/cluster.py]: `.format` not needed

File: sdcm/cluster.py [hunk 13]
```diff
@@ -2134,11 +2300,34 @@ def wait_for_init(self):
...
+            prereqs_script = dedent("""
+                curl -fsSL https://download.docker.com/linux/debian/gpg | sudo apt-key add -
+                sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable"
+                sudo apt-get update
+                sudo apt-get install -y docker docker.io
+                apt-get install -y python-setuptools unzip wget
+                easy_install pip
+                pip install --upgrade pip
+                pip install pyyaml
+            """.format(**locals()))
```
Comment (bentsi) [inline sdcm/cluster.py]: same here: `.format` not needed

File: sdcm/collectd.py [hunk 14]
```diff
@@ -426,19 +433,49 @@ def collectd_exporter_setup(self):
...
-                self.node.remoter.run('sudo systemctl start collectd-exporter.service')
+                if self.node.is_rhel_like() or self.node.is_ubuntu16():
+                    self.node.remoter.run('sudo systemctl start collectd-exporter.service')
+                else:
+                    self.node.remoter.run('sudo service collectd-exporter start')
+        finally:
+            shutil.rmtree(tmp_dir_exporter)
+
+    def collectd_exporter_service_setup(self):
+        service_file = """# Run node_exporter
```
Comment (bentsi) [inline sdcm/collectd.py]: Use dedent("...") like above

**Comment by roydahan** [comment]
@amoskong LGTM, please update/answer comments from Bentsi.

**Comment by roydahan** [comment]
@amoskong / @bentsi , what's blocking this from getting in?

**Comment by amoskong** [comment]
I didn't address the final comment, will do it and rebase to latest master.

**Comment by roydahan** [comment]
let's merge.
